src/hdmf/validate/validator.py

```diff
@@ -339,28 +356,44 @@ class BaseStorageValidator(Validator):
             {'name': 'validator_map', 'type': ValidatorMap, 'doc': 'the ValidatorMap to use during validation'})
     def __init__(self, **kwargs):
         call_docval_func(super().__init__, kwargs)
-        self.__attribute_validators = dict()
-        for attr in self.spec.attributes:
-            self.__attribute_validators[attr.name] = AttributeValidator(attr, self.vmap)
 
     @docval({"name": "builder", "type": BaseBuilder, "doc": "the builder to validate"},
             returns='a list of Errors', rtype=list)
     def validate(self, **kwargs):
         builder = getargs('builder', kwargs)
-        attributes = builder.attributes
-        ret = list()
-        for attr, validator in self.__attribute_validators.items():
-            attr_val = attributes.get(attr)
-            if attr_val is None:
-                if validator.spec.required:
-                    ret.append(MissingError(self.get_spec_loc(validator.spec),
-                                            location=self.get_builder_loc(builder)))
+        errors = list(self.__validate_attributes(builder))
+        return errors
+
+    def __validate_attributes(self, builder):
```

Comment:
Instead of comments, I refactored out new functions and added clarification in the docstrings. Please let me know if you think some comments would still be beneficial here, happy to add if so!